File: src/002.String/541.ReverseStringII/541.反转字符串-ii.js
/*
 * @lc app=leetcode.cn id=541 lang=javascript
 *
 * [541] 反转字符串 II
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var reverseStr = function (s, k) {
	const n = s.length;
	const arr = Array.from(s);
	for (let i = 0; i < n; i += 2 * k) {
		reverse(arr, i, Math.min(i + k - 1, n - 1));
	}
	return arr.join('');
};

function reverse(arr, left, right) {
	while (left < right) {
		[arr[left], arr[right]] = [arr[right], arr[left]];
		left++;
		right--;
	}
}
// @lc code=end

File: src/002.String/541.ReverseStringII/541.反转字符串-ii.go
/*
 * @lc app=leetcode.cn id=541 lang=golang
 *
 * [541] 反转字符串 II
 */

// @lc code=start
func reverseStr(s string, k int) string {
	arr := []byte(s)
	for i := 0; i < len(s); i += 2 * k {
		end := min(i+k-1, len(s)-1)
		reverse(arr, i, end)
	}
	return string(arr)
}

func min(a int, b int) int {
	if a > b {
		return b
	}
	return a
}

func reverse(arr []byte, left int, right int) {
	for left < right {
		arr[left], arr[right] = arr[right], arr[left]
		left++
		right--
	}
}

// @lc code=end

